File: solutions/go/bob/2/bob.go
// Package bob determine what Bob will reply to someone when they say something to him or ask him a question
package bob

import (
	"regexp"
	"strings"
)

// Hey should determine what Bob will reply to a remark.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)
	if remark == "" {
		return "Fine. Be that way!"
	}

	hasCapitalLetters := regexp.MustCompile(`[A-Z]`).MatchString(remark)
	hasDowncaseLetters := regexp.MustCompile(`[a-z]`).MatchString(remark)
	hasOnlyCapitalLetters := hasCapitalLetters && !hasDowncaseLetters
	isQuestion := strings.HasSuffix(remark, "?")

	switch true {
	case isQuestion && hasOnlyCapitalLetters:
		return "Calm down, I know what I'm doing!"
	case isQuestion:
		return "Sure."
	case hasOnlyCapitalLetters:
		return "Whoa, chill out!"
	default:
		return "Whatever."
	}
}

File: solutions/elixir/bob/2/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input), do: input |> String.trim() |> reply()

  @spec reply(String.t()) :: String.t()
  def reply(""), do: "Fine. Be that way!"
  def reply(input) do
    is_question = String.ends_with?(input, "?")
    is_yelled = input === String.upcase(input) and Regex.match?(~r/\p{Lu}/u, input)

    cond do
      is_question and is_yelled ->
        "Calm down, I know what I'm doing!"

      is_question ->
        "Sure."

      is_yelled ->
        "Whoa, chill out!"

      true ->
        "Whatever."
    end
  end
end
